# Practico 1 - Aprendizaje Profundo

Alumnos:
* Milagros Avila
* Juan Porta
* Ignacio Scarinci

Se hicieron dos pipeline: uno para el ajuste de los distintos modelos **practico_1_train_petfinder_modelos.py** y otro para la busqueda de hiperparametros **practico_1_train_petfinder_busqueda.py**

In [2]:
import pandas as pd

In [3]:
res_mod_1 = pd.read_csv('./resultados_modelo1.csv')
res_mod_2 = pd.read_csv('./resultados_modelo2.csv')
res_mod_3 = pd.read_csv('./resultados_modelo3.csv')
res_mod_4 = pd.read_csv('./resultados_modelo4.csv')


Se probaron 4 modelos que se pueden ver en la siguiente figura:

**Modelo 1:**
![Modelo 1](modelo_1.png)

**Modelo 2:**
![Modelo 2](modelo_2.png)

**Modelo 3:**
![Modelo 3](modelo_3.png)

**Modelo 4:**
![Modelo 4](modelo_4.png)

Implementación de los modelos en pyhton

```python

modelo = args.modelo #Se pasa como argumento al pipeline el número de modelo a experimentar

if modelo == 1:
    dense1 = layers.Dense(100,activation='relu')(features)
    dense_ = layers.Dense(100,activation='relu')(dense1)
elif modelo == 2:
    bn = layers.BatchNormalization(momentum=0)(features)
    dense1 = layers.Dense(80,activation='relu')(bn)
    dropout_1 = layers.Dropout(0.5)(dense1)
    dense2 = layers.Dense(100,activation='relu')(dropout_1)
    dense_ = layers.Dense(100,activation='relu')(dense2)
elif modelo == 3:
    dense1 = layers.Dense(80,activation='relu')(features)
    dense2 = layers.Dense(100,activation='relu')(dense1)
    dropout_1 = layers.Dropout(0.5)(dense2)
    dense_ = layers.Dense(100,activation='relu')(dropout_1)
elif modelo == 4:
    bn = layers.BatchNormalization(momentum=0)(features)
    dense1 = layers.Dense(80,activation='relu')(bn)
    dense2 = layers.Dense(90,activation='relu')(dense1)
    dropout_1 = layers.Dropout(0.2)(dense2)
    dense2 = layers.Dense(90,activation='relu')(dropout_1)
    dropout_2 = layers.Dropout(0.2)(dense2)
    dense_ = layers.Dense(90,activation='relu')(dropout_2)

output_layer = layers.Dense(nlabels, activation='softmax')(dense_)

model = models.Model(inputs=inputs, outputs=output_layer)

optimizador = tf.keras.optimizers.Adam(learning_rate=1e-4)
model.compile(loss='categorical_crossentropy', optimizer=optimizador,
              metrics=['accuracy'])
```

Además se probaron 3 combinaciones de feactures para cada modelo.

In [4]:
print('Combinación 1:')

print('Embedded columns =', res_mod_1['embedded_columns'][2])
print('One hot columns =', res_mod_1['one_hot_columns'][2])
print('Numerical columns =', res_mod_1['numerical_columns'][2])

print('\nCombinación 2:')

print('Embedded columns =', res_mod_1['embedded_columns'][1])
print('One hot columns =', res_mod_1['one_hot_columns'][1])
print('Numerical columns =', res_mod_1['numerical_columns'][1])

print('\nCombinación 3:')

print('Embedded columns =', res_mod_1['embedded_columns'][0])
print('One hot columns =', res_mod_1['one_hot_columns'][0])
print('Numerical columns =', res_mod_1['numerical_columns'][0])

Combinación 1:
Embedded columns = {'Breed1': 308}
One hot columns = {'Gender': 3, 'Color1': 7}
Numerical columns = ['Age', 'Fee']

Combinación 2:
Embedded columns = {'Breed1': 308, 'Breed2': 308}
One hot columns = {'Gender': 3, 'Color1': 7, 'Type': 2, 'MaturitySize': 4, 'Sterilized': 3}
Numerical columns = ['Age', 'Fee']

Combinación 3:
Embedded columns = {'Breed1': 308, 'Breed2': 308}
One hot columns = {'Gender': 3, 'Color1': 7, 'Color2': 7, 'Color3': 7, 'Type': 2, 'MaturitySize': 4, 'Sterilized': 3, 'Vaccinated': 3, 'Health': 3}
Numerical columns = ['Age', 'Fee']


Resultados obtenidos para cada modelo y combinacón de feactures

In [5]:
for i in range(2,-1,-1):
    print('{} \n Accuracy:{} \n Accuracy Validation:{} \n Loss:{} \n Validation Loss:{} \n'.format(res_mod_1['Name'][i],
                                                                                            res_mod_1['accuracy'][i],
                                                                                            res_mod_1['val_accuracy'][i], 
                                                                                            res_mod_1['loss'][i],
                                                                                            res_mod_1['val_loss'][i]))
print('========================================')
    
for i in range(2,-1,-1):
    print('{} \n Accuracy:{} \n Accuracy Validation:{} \n Loss:{} \n Validation Loss:{} \n'.format(res_mod_2['Name'][i],
                                                                                            res_mod_2['accuracy'][i],
                                                                                            res_mod_2['val_accuracy'][i], 
                                                                                            res_mod_2['loss'][i],
                                                                                            res_mod_2['val_loss'][i]))
print('========================================')
for i in range(2,-1,-1):
    print('{} \n Accuracy:{} \n Accuracy Validation:{} \n Loss:{} \n Validation Loss:{} \n'.format(res_mod_3['Name'][i],
                                                                                            res_mod_3['accuracy'][i],
                                                                                            res_mod_3['val_accuracy'][i], 
                                                                                            res_mod_3['loss'][i],
                                                                                            res_mod_3['val_loss'][i]))
print('========================================')
for i in range(2,-1,-1):
    print('{} \n Accuracy:{} \n Accuracy Validation:{} \n Loss:{} \n Validation Loss:{} \n'.format(res_mod_4['Name'][i],
                                                                                            res_mod_4['accuracy'][i],
                                                                                            res_mod_4['val_accuracy'][i], 
                                                                                            res_mod_4['loss'][i],
                                                                                            res_mod_4['val_loss'][i]))

Modelo 1 Combinación 1 
 Accuracy:0.35369167 
 Accuracy Validation:0.30751059999999997 
 Loss:1.4103189689199107 
 Validation Loss:1.4462323758139541 

Modelo 1 Combinación 2 
 Accuracy:0.39586532 
 Accuracy Validation:0.36183277 
 Loss:1.3539928544433737 
 Validation Loss:1.40839139027382 

Modelo 1 Combinación 3 
 Accuracy:0.40732428 
 Accuracy Validation:0.3490789 
 Loss:1.340043658838424 
 Validation Loss:1.411627228580304 

Modelo 2 Combinación 1 
 Accuracy:0.33585352 
 Accuracy Validation:0.31743032 
 Loss:1.4208604594059968 
 Validation Loss:1.4673980421094752 

Modelo 2 Combinación 2 
 Accuracy:0.38133490000000003 
 Accuracy Validation:0.34246576 
 Loss:1.3703482021753137 
 Validation Loss:1.4340022072863223 

Modelo 2 Combinación 3 
 Accuracy:0.37991729999999996 
 Accuracy Validation:0.35144073 
 Loss:1.365807772795265 
 Validation Loss:1.4579491223861922 

Modelo 3 Combinación 1 
 Accuracy:0.33479032 
 Accuracy Validation:0.322154 
 Loss:1.4278766784735566 
 Validation Loss:1

Las siguientes imágenes muestran en un grafico paralelo el efecto de agregar features 

**Modelo 1**
![Modelo1_features](modelo_1_features.png)

**Modelo 2**
![Modelo2_features](modelo_2_features.png)

**Modelo 3**
![Modelo3_features](modelo_3_features.png)

**Modelo 4**
![Modelo4_features](modelo_4_features.png)

Como se puede ver en los modelos 1, 3 y 4 cuando se agregan feactures mejora el accuracy y baja la perdida. En el modelo 2 no ocurre este fenomeno. 

Dado que mejora en la mayoria de los casos el accuracy se deside realizar la bsqueda de los hiperparametros optimos utilizando la combinacion de feactures denominada 3.

### Hiperparametros optimos

Se realizo una bsqueda Bayesiana del valor optimo para el númeoro de neuronas para cada capa, el valor de dropout para cada capa de dropout, la tasa de aprendizaje y el número de epocas.

Para la busqueda de los valores optimos se utilizo como parametro el accuracy *(-1)

Dado que se buscaba el menor accuracy en algunos casos cuando se usaban los parametros optimos no se optenian buenos resultados se muestran los resulados para los parametros con mejor accuracy durante la busqueda y los de menor perdida.

#### Modelo 1:

**Mejor accuracy:** 
*  epochs:25
*  hidden_layer_1:65
*  hidden_layer_2:65
*  learning_rate:0.0001

<div>
<img src="modelo_1_optimo_acc_acc.png" width="500"/> 
<img src="modelo_1_optimo_loss_acc.png" width="500"/>
</div>


Resultados:
*  accuracy:0.405
*  loss:1.34
*  val_accuracy:0.355
*  val_loss:1.412


**Menor loss:**
*  epochs:25
*  hidden_layer_1:130
*  hidden_layer_2:65
*  learning_rate:0.0001


<div>
<img src="modelo_1_optimo_acc_loss.png" width="500"/> 
<img src="modelo_1_optimo_loss_loss.png" width="500"/>
</div>

Resultados
*  accuracy:0.41
*  loss:1.324
*  val_accuracy:0.362
*  val_loss:1.394


#### Modelo 2:

**Mejor accuracy:**
    
*  epochs: 35
*  hidden_layer_1: 72
*  hidden_layer_2: 125
*  hidden_layer_3: 83
*  dropout_1: 0.486
*  learning_rate:0.000978

<div>
<img src="modelo_2_optimo_acc_acc.png" width="500"/> 
<img src="modelo_2_optimo_loss_acc.png" width="500"/>
</div>


Resultados:
*  accuracy:0.44
*  loss:1.233
*  val_accuracy:0.357
*  val_loss:1.713


**Menor loss:**
*  epochs:25
*  hidden_layer_1:65
*  hidden_layer_2:65
*  hidden_layer_2:130
*  dropout_1: 0.499
*  learning_rate:0.0001


<div>
<img src="modelo_2_optimo_acc_loss.png" width="500"/> 
<img src="modelo_2_optimo_loss_loss.png" width="500"/>
</div>

Resultados
*  accuracy:0.394
*  loss:1.357
*  val_accuracy:0.332
*  val_loss:1.527





#### Modelo 3:

**Mejor accuracy:**
    
*  epochs: 25
*  hidden_layer_1: 84
*  hidden_layer_2: 130
*  hidden_layer_3: 65
*  dropout_1: 0.4999
*  learning_rate:0.0001

<div>
<img src="modelo_3_optimo_acc_acc.png" width="500"/> 
<img src="modelo_3_optimo_loss_acc.png" width="500"/>
</div>


Resultados:
*  accuracy:0.382
*  loss:1.364
*  val_accuracy:0.366
*  val_loss:1.401


**Menor loss:**
*  epochs:32
*  hidden_layer_1:65
*  hidden_layer_2:130
*  hidden_layer_2:130
*  dropout_1: 0.01
*  learning_rate:0.0001


<div>
<img src="modelo_3_optimo_acc_loss.png" width="500"/> 
<img src="modelo_3_optimo_loss_loss.png" width="500"/>
</div>

Resultados
*  accuracy:0.432
*  loss:1.292
*  val_accuracy:0.347
*  val_loss:1.425


#### Modelo 4:

**Mejor accuracy:**
    
*  epochs: 25
*  hidden_layer_1: 130
*  hidden_layer_2: 65
*  hidden_layer_3: 130
*  hidden_layer_4: 130
*  hidden_layer_5: 130
*  dropout_1: 0.4999
*  dropout_1: 0.01
*  learning_rate:0.0001

<div>
<img src="modelo_4_optimo_acc_acc.png" width="500"/> 
<img src="modelo_4_optimo_loss_acc.png" width="500"/>
</div>


Resultados:
*  accuracy: 0.4
*  loss: 1.377
*  val_accuracy: 0.357
*  val_loss: 1.524


**Menor loss:**

*  epochs: 25
*  hidden_layer_1: 130
*  hidden_layer_2: 65
*  hidden_layer_3: 130
*  hidden_layer_4: 130
*  hidden_layer_5: 130
*  dropout_1: 0.4999
*  dropout_1: 0.4999
*  learning_rate:0.0001

<div>
<img src="modelo_4_optimo_acc_loss.png" width="500"/> 
<img src="modelo_4_optimo_loss_loss.png" width="500"/>
</div>

Resultados
*  accuracy:0.376
*  loss:1.406
*  val_accuracy:0.359
*  val_loss:1.505


### Conclusiones 

* 3 capas sin regularizacion o dropout se overfitiaba
* El modelo que da mejores resultados es el modelo 3 